### PR TITLE
fix(theming): Conitionally disable blur filter for performance

### LINKED ISSUE #7896: CSS blur filter seriously impacts performance in Chrome
Notice the sharp fps drop when the filter is active

https://user-images.githubusercontent.com/26852655/189666173-71eda72e-24c4-419d-8a43-2ae4ce7bb900.mov

### LINKED ISSUE #44122: scrolling in dashboard is laggy
just tested on nc29.0.0b1
Scrolling in dashboard app is laggy in chromium based browser:

https://github.com/nextcloud/server/assets/42591237/cc20a0f0-4a9b-46b9-ab52-9da7a0445c20

### PR DESCRIPTION
**Pull request by susnux**
* Resolves: https://github.com/nextcloud/spreed/issues/7896
* Closes: https://github.com/nextcloud-libraries/nextcloud-vue/pull/5607
* Closes: #45049
* Resolves: https://github.com/nextcloud/server/issues/44122

## Summary

As an alternative for #45049 but this one is not removing the blurry background for all users but only for chrome + edge by default.
User can force enable if it works for them (e.g. they have hardware acceleration) or force disable if they suffer performance issues.

Firefox and Safari are not affected by the performance issues, and Android should also be ok as they have proper hardware acceleration.
Mostly Windows is affected and Linux with faulty GPU drivers.

### Screencast

https://github.com/nextcloud/server/assets/1855448/bdd05a79-69d8-4435-bfdf-73290fb253b9

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/theming/lib/Themes/DefaultTheme.php [hunk 1]
```diff
@@ -120,12 +107,23 @@ public function getCSSVariables(): array {
 		$colorSuccess = '#2d7b41';
 		$colorInfo = '#0071ad';
 
+		// Chromium based browsers currently (2024) have huge performance issues with blur filters
+		$isChromium = $this->request !== null && $this->request->isUserAgent([Request::USER_AGENT_CHROME, Request::USER_AGENT_MS_EDGE]);
```
Comment (solracsf) [inline apps/theming/lib/Themes/DefaultTheme.php:111]: Maybe also add mobile variant?

```suggestion
		$isChromium = $this->request !== null && $this->request->isUserAgent([Request::USER_AGENT_CHROME, Request::USER_AGENT_MS_EDGE, Request::USER_AGENT_ANDROID_MOBILE_CHROME]);
```

Comment (szaimen) [reply apps/theming/lib/Themes/DefaultTheme.php:111]: Just tested and on android it does not suffer performance problems

**Comment by susnux** [comment]
To be discussed by @nextcloud/designers

File: apps/theming/lib/Themes/DefaultTheme.php (hunk 1 above)
Comment (susnux) [reply apps/theming/lib/Themes/DefaultTheme.php:111]: Android should have proper hardware acceleration, is is mostly a chromium + GPU driver issue.

**Comment by kesselb** [comment]
Thanks for working on it :+1: 

Apparently my chromium has no hardware acceleration any longer  (it's a snap package now, maybe that's related). 

It's basically impossible to use Nextcloud and especially navigation and dashboard.

[Screencast from 2024-05-18 16-46-36.webm](https://github.com/nextcloud/server/assets/3902676/732a7194-702e-4a2c-ad56-ae0109f4e689)

[Screencast from 2024-05-18 16-55-12.webm](https://github.com/nextcloud/server/assets/3902676/16621166-f5d2-4bf2-b081-b7205496c49b)

The auto-detection worked. 
If you selected yes or no once, you cannot go back to "auto" but that's acceptable, I guess. 
The left navigation looks definitely cleaner with the blurred background.

**Review by marcoambrosini — DISMISSED** [review]
I like this approach in general :) 
One thing: leaving the sharp edges of a picture below the text can make it hard to read. I think that when we disable the blur we should also add a solid background, not transparent. We can use an average of the background similar to the auto primary? But lighter or darker depending on the theme.
Can we also narrow this down to windows && chromium users only?

**Comment by susnux** [comment]
> windows && chromium

It is not only Windows, I once got this (I run a rolling release distro) on Linux and had to wait like 2 days for bleeding edge Chromium fixing their GPU layer (not sure what they are using because I never had issues with Firefox).

>  One thing: leaving the sharp edges of a picture below the text can make it hard to read. I think that when we disable the blur we should also add a solid background, not transparent. We can use an average of the background similar to the auto primary? But lighter or darker depending on the theme.

For 30+ with #42977 this would work fine, but with older (we need to backport) we can basically only mix with primary color and hope it matches somewhat the background.

**Comment by marcoambrosini** [comment]
>  I once got this (I run a rolling release distro) on Linux and had to wait like 2 days

I think for such a case we can reasonably expect people to go and find the checkbox themselves

Another Idea that I had in the past was creating a blurred version of the background on the server and use that one with `position: fixed` behind all transparent elements. What do you think @susnux?

**Comment by kesselb** [comment]
> I think for such a case we can reasonably expect people to go and find the checkbox themselves

If chromium/chrome can use the hardware acceleration depends on the operating system, gpu, gpu drivers and packaging. We cannot just assume it does not work on Windows but everywhere else. I'm using Ubuntu 24.04 with Chromium (snap), and have no hardware acceleration either. It did work with Ubuntu 23.10 and Chromium (apt). 

image

I think you should get an idea, how annoying the current state is, if you turn off the graphics acceleration in your chrome/chromium settings.

**Comment by marcoambrosini** [comment]
> I'm using Ubuntu 24.04 with Chromium 

That's the point, it seems like a very niche use case. People with similar setups can surely reach into the settings and force-disable the blur. 

That said, what do you think about [this idea](https://github.com/nextcloud/server/pull/45395#issuecomment-2119730795)
also cc @juliushaertl and @ShGKme

**Comment by kesselb** [comment]
> That's the point, it seems like a very niche use case compared to the vast majority of chrome based browsers on platforms other than windows. People with setups similar to yours can surely reach into the settings and force-disable the blur.

How would someone know that for the sluggish appearance is caused by a blur filter that can be turned off by a configuration option? ;) 

What are your thoughts on enabling the blur filter by default only for chrome/chromium on macosx and android?


> That said, what do you think about https://github.com/nextcloud/server/pull/45395#issuecomment-2119730795

I am not sure if I understood the idea. Server will provide a blurred version of the background image. The default background is the not-blurred version. The blurred version is placed on top of the default background image and only visible if there is a transparent element to be shown? Is the background image then loaded twice, and how does an element know that the other background is needed?

**Comment by susnux** [comment]
> That said, what do you think about [this idea](https://github.com/nextcloud/server/pull/45395#issuecomment-2119730795)

I think this is hard to be done correctly implementation-wise, meaning to ensure the blurred version is shown on the correct location with correct scaling.
As said with Nextcloud 30 we can have a fake blurhash that we can use there, for 28 and 29 I think we could instead just mix background and primary.

This would need no changes at all in all apps and library, we just need to patch server theming variables.

For example:

With blur | Without blur | With mixed-color
---|---|---
Screenshot_20240520_200400|Screenshot_20240520_200355|Screenshot_20240520_200543

**Comment by marcoambrosini** [comment]
@susnux I fiddled a bit with it and I think we can accomplish this with a before element. Here's a code pen:

https://codepen.io/marukodesu/pen/RwmRbpr

If this can be solved without creating a setting, I think we should go down that route.

**Comment by ShGKme** [comment]
> Another Idea that I had in the past was creating a blurred version of the background on the server

We can also start with creating it on frontend side once on load, but could be not very fast as well.

**Comment by marcoambrosini** [comment]
I created [an issue with the approach I propose](https://github.com/nextcloud/server/issues/45452)  for 30+

**Comment by jancborchardt** [comment]
@marcoambrosini your suggestion in https://github.com/nextcloud/server/issues/45452 sounds good to me – better than any setting. What do you think @ShGKme @susnux @kesselb?

As @kesselb said:

> How would someone know that for the sluggish appearance is caused by a blur filter that can be turned off by a configuration option? ;)

So a config option is not it. But outright disabling it for a bunch of people is also not nice since then it will look way polished. The 2-image option makes it look the same by fixing it in the background – neat. :)

**Review by jancborchardt — COMMENTED** [review]
@marcoambrosini your suggestion in https://github.com/nextcloud/server/issues/45452 sounds good to me – better than any setting. What do you think @ShGKme @susnux @kesselb?

As @kesselb said:

> How would someone know that for the sluggish appearance is caused by a blur filter that can be turned off by a configuration option? ;)

So a config option is not it. But outright disabling it for a bunch of people is also not nice since then it will look way polished. The 2-image option makes it look the same by fixing it in the background – neat. :)

**Comment by netzpolitikorg** [comment]
> I think for such a case we can reasonably expect people to go and find the checkbox themselves

if they can find it? we're on 29.0.3 and can't see that option :disappointed:

**Comment by susnux** [comment]
@marcoambrosini Mac OS is now ignored because it always have hardware acceleration

**Review by jancborchardt — COMMENTED** [review]
@susnux with this pull request, most people will not see the visual effect of blurring (however it is achieved) anymore, as far as I understand? Which browsers is it default enabled and which default disabled for?

I was under the impression we would go for https://github.com/nextcloud/server/issues/45452 as an inbetween solution instead. cc @AndyScherzinger since removing blur will make stuff look less nice.

**Comment by marcoambrosini** [comment]
If it's impossible to find a solution I would remove transparency + blur everywhere and work from there to achieve a nice design for everyone, without settings

**Comment by susnux** [comment]
> I was under the impression we would go for #45452 as an inbetween solution instead. cc @AndyScherzinger since removing blur will make stuff look less nice.

Yes I hope we get this for Nextcloud 30. But we need a backportable solution for 28 and 29 as currently on those machines you can not use Nextcloud.
If you prefer I remove the autodetect part and only add the option switch. 

For 30 we can then use @marcoambrosini solution

**Comment by susnux** [comment]
> Which browsers is it default enabled and which default disabled for?

Currently in this PR:
✅ everything not Chromium based
✅ Chromium based on mobile
✅ Chromium based on MacOS
❌ Chromium based on Windows
❌ Chromium based on Linux 

The first 3 never suffer the issue 
For Windows and Linux it depends on your GPU driver

**Comment by marcoambrosini** [comment]
> we need a backportable solution

Got it 

> If you prefer I remove the autodetect part and only add the option switch

I think it's fine then to leave the autodetect on

File: apps/theming/lib/Themes/DefaultTheme.php [hunk 2]
```diff
@@ -96,12 +83,28 @@ public function getCSSVariables(): array {
...
+		// Ignore MacOS because they always have hardware accelartion
+		$isChromium = $isChromium && !$this->request->isUserAgent(['/Macintosh/']);
+		// Allow to force the blur filter
+		$forceEnableBlur = $this->config->getUserValue(
+			$this->userSession->getUser()->getUID(),
```
Comment (szaimen) [inline apps/theming/lib/Themes/DefaultTheme.php:92]: I juest tested this and got `Call to a member function getUID() on null in file '/var/www/nextcloud/apps/theming/lib/Themes/DefaultTheme.php' line 92` in the logs

**Review by jancborchardt — APPROVED** [review]
Thanks for the explanation @susnux in your comment https://github.com/nextcloud/server/pull/45395#issuecomment-2201470751

Then I agree with you we should get this in very soon to backport, and get Marco’s solution done for 30: https://github.com/nextcloud/server/issues/45452

@sorbaugh

File: apps/theming/lib/Themes/DefaultTheme.php (hunk 2 above)
Comment (szaimen) [reply apps/theming/lib/Themes/DefaultTheme.php:92]: Yes public page. I'd vote for 1. in that case

Comment (susnux) [reply apps/theming/lib/Themes/DefaultTheme.php]: fixed

**Review by szaimen — APPROVED** [review]
Tested and works :)

However it is a bit weird to show this state by default which will first allow to disable it even though it is already disabled: 
image

**Comment by szaimen** [comment]
🎉🎉🎉🎉🎉

**Comment by szaimen** [comment]
> > Which browsers is it default enabled and which default disabled for?
> 
> Currently in this PR: ✅ everything not Chromium based ✅ Chromium based on mobile ✅ Chromium based on MacOS ❌ Chromium based on Windows ❌ Chromium based on Linux
> 
> The first 3 never suffer the issue For Windows and Linux it depends on your GPU driver

@susnux I guess we should document this somewhere?

**Comment by szaimen** [comment]
I suppose on windows and linux with chromium based browser the checkbox should also be disabled? Apparently @peterwebdesign ran into a weird bug with it...

**Pull request merged**